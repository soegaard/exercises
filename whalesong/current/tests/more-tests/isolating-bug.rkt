#lang planet dyoo/whalesong/base
(require (for-syntax racket/base
                     syntax/struct))


(define-syntax (my-define-struct stx)
  (syntax-case stx ()
    [(_ name (fields ...) kw ...)
     (with-syntax ([(names ...)
                    (build-struct-names #'name
                                        (syntax->list #'(fields ...))
                                        #f
                                        #f)])
       (with-syntax ([cnstr (syntax-case #'(names ...) ()
                              [(struct:name-id constructor misc ...)
                               #'constructor])])
         #'(begin
             (define-values (names ...)
               (let ()
                 (define-struct name (fields ...) kw ...)
                 (call-with-values (lambda ()
                                     (let ([cnstr (lambda args
                                                    (apply cnstr args))])
                                       (displayln names) ...
                                       (values names ...)))
                                   (lambda args
                                     (displayln "in the result of call-with-values")
                                     (displayln args)
                                     (apply values args))))))))]))







(define-syntax (my-define-struct2 stx)
  (syntax-case stx ()
    [(_ name (fields ...) kw ...)
     (with-syntax ([(names ...)
                    (build-struct-names #'name
                                        (syntax->list #'(fields ...))
                                        #f
                                        #f)])
       (with-syntax ([cnstr (syntax-case #'(names ...) ()
                              [(struct:name-id constructor misc ...)
                               #'constructor])])
         #'(begin
              (define-values (names ...)
                (let ()
                 (begin
                   (define-struct name (fields ...) kw ...)
                   (let ([cnstr (lambda args
                                 (apply cnstr args))])
                     (displayln names) ...
                     (values names ...))))))))]))



(my-define-struct swf (f) #:mutable)
(displayln "---")
struct:swf
make-swf
swf?
swf-f
set-swf-f!


(displayln "***")


(my-define-struct swf2 (f) #:mutable)
(displayln "---")
struct:swf2
make-swf2
swf2?
swf2-f
set-swf2-f!
