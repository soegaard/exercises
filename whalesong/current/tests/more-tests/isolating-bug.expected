#<struct-type:swf>
#<procedure:make-swf>
#<procedure:swf?>
#<procedure:swf-f>
#<procedure:set-swf-f!>
in the result of call-with-values
(#<struct-type:swf> #<procedure:make-swf> #<procedure:swf?> #<procedure:swf-f> #<procedure:set-swf-f!>)
---
#<struct-type:swf>
#<procedure:make-swf>
#<procedure:swf?>
#<procedure:swf-f>
#<procedure:set-swf-f!>
***
#<struct-type:swf2>
#<procedure:make-swf2>
#<procedure:swf2?>
#<procedure:swf2-f>
#<procedure:set-swf2-f!>
in the result of call-with-values
(#<struct-type:swf2> #<procedure:make-swf2> #<procedure:swf2?> #<procedure:swf2-f> #<procedure:set-swf2-f!>)
---
#<struct-type:swf2>
#<procedure:make-swf2>
#<procedure:swf2?>
#<procedure:swf2-f>
#<procedure:set-swf2-f!>
